src/t9p_atomic.h

```diff
@@ -0,0 +1,133 @@
+    return *p; \
+  }
+
+#define T9P__DEFINE_ATOMIC_STORE(shorthand, type) \
+  static inline type \
+  t9p_atomic_store_##shorthand(type* p, type val) \
+  { \
+    ASSERT_ALIGNED(p, sizeof(type)); \
+    *p = val; \
+  }
+
+#define T9P__DEFINE_ATOMIC_COMPARE_EXCHANGE(shorthand, type) \
+  static inline int \
+  t9p_atomic_compare_exchange_##shorthand(type* p, type expected, type newval) \
+  { \
+    ASSERT_ALIGNED(p, sizeof(type)); \
+    rtems_interrupt_level l; \
+    int r = 0; \
+    rtems_interrupt_disable(l); \
+    if (*p == expected) \
+      r = 1, *p = newval; \
+    rtems_interrupt_enable(l); \
+    return r; \
+  }
+
+#define T9P__DEFINE_ATOMIC_ADD(shorthand, type) \
+  static inline type \
+  t9p_atomic_add_##shorthand(type* p, type v) \
+  { \
+    ASSERT_ALIGNED(p, sizeof(type)); \
+    type r; \
+    rtems_interrupt_level l; \
+    rtems_interrupt_disable(l); \
+    (*p) += v; \
+    r = *p; \
+    rtems_interrupt_enable(l); \
+    return r; \
+  }
+
+#define T9P__DEFINE_ATOMIC_SUB(shorthand, type) \
+  static inline type \
+  t9p_atomic_sub_##shorthand(type* p, type v) \
+  { \
+    ASSERT_ALIGNED(p, sizeof(type)); \
+    type r; \
+    rtems_interrupt_level l; \
+    rtems_interrupt_disable(l); \
+    (*p) -= v; \
+    r = *p; \
+    rtems_interrupt_enable(l); \
+    return r; \
+  }
+
+#else
+
+#define T9P__DEFINE_ATOMIC_LOAD(shorthand, type) \
+  static inline type \
+  t9p_atomic_load_##shorthand(type* p) \
+  { \
+    return __atomic_load_n(p, __ATOMIC_SEQ_CST); \
+  }
+
+#define T9P__DEFINE_ATOMIC_STORE(shorthand, type) \
+  static inline type \
+  t9p_atomic_store_##shorthand(type* p, type val) \
+  { \
+    ASSERT_ALIGNED(p, sizeof(type)); \
+    __atomic_store_n(p, val, __ATOMIC_SEQ_CST); \
+  }
+
+#define T9P__DEFINE_ATOMIC_COMPARE_EXCHANGE(shorthand, type) \
+  static inline int \
+  t9p_atomic_compare_exchange_##shorthand(type* p, type expected, type newval) \
+  { \
+    ASSERT_ALIGNED(p, sizeof(type)); \
+    return __atomic_compare_exchange( \
+      p, &expected, &newval, 0, __ATOMIC_SEQ_CST, __ATOMIC_RELAXED \
+    ); \
+  }
+
+#define T9P__DEFINE_ATOMIC_ADD(shorthand, type) \
+  static inline type \
+  t9p_atomic_add_##shorthand(type* p, type v) \
+  { \
+    ASSERT_ALIGNED(p, sizeof(type)); \
+    return __atomic_add_fetch(p, v, __ATOMIC_SEQ_CST); \
+  }
+
+#define T9P__DEFINE_ATOMIC_SUB(shorthand, type) \
+  static inline type \
+  t9p_atomic_sub_##shorthand(type* p, type v) \
+  { \
+    ASSERT_ALIGNED(p, sizeof(type)); \
+    return __atomic_sub_fetch(p, v, __ATOMIC_SEQ_CST); \
+  }
+
+#endif
+
+#define T9P__DEFINE_ATOMIC_OPERATIONS(shorthand, type) \
+   T9P__DEFINE_ATOMIC_LOAD(shorthand, type) \
+   T9P__DEFINE_ATOMIC_STORE(shorthand, type) \
+   T9P__DEFINE_ATOMIC_COMPARE_EXCHANGE(shorthand, type) \
+   T9P__DEFINE_ATOMIC_ADD(shorthand, type) \
+   T9P__DEFINE_ATOMIC_SUB(shorthand, type)
+
+T9P__DEFINE_ATOMIC_OPERATIONS(u32, uint32_t)
+T9P__DEFINE_ATOMIC_OPERATIONS(u64, uint64_t)
```

Comment:
32-bit PowerPC is missing most of the `__atomic` builtins on GCC, they generate linker errors. They'll need to be implemented in the same way as the Coldfire atomics using `rtems_interrupt_enable`/`disable`. It's probably best to leave these as manually implemented inlines (as they were before), as ugly as that may be...

Reply:
when building with rtems4-beatnik:
```
CMakeFiles/t9p_rtems_test.dir/tests/t9p_cmd.c.obj: In function `t9p_atomic_load_u64':
/home/jeremy/dev/lw9p/src/t9p_atomic.h:133: undefined reference to `__atomic_load_8'
libt9p.a(t9p_mem.c.obj): In function `t9p_atomic_add_u64':
/home/jeremy/dev/lw9p/src/t9p_atomic.h:133: undefined reference to `__atomic_fetch_add_8'
```